##Assignment 2 2008160141 임길환

###•다음 6 개의 Expression에 대해 Evaluation 결과 값을 출력하고, 해당 결과가 나온 이유에 대해 설명하시오
◾1 and 2 and 3 and 4
◾1 or 2 or 3 or 4
◾1 and 2 or 3 and 4
◾(1 and 2) or (3 and 4)
◾1 or 2 and 3 or 4
◾(1 or 2) and (3 or 4)


In [22]:
print 1 and 2 and 3 and 4
print 1 or 2 or 3 or 4
print 1 and 2 or 3 and 4
print (1 and 2) or (3 and 4)
print 1 or 2 and 3 or 4
print (1 or 2) and (3 or 4)

4
1
2
2
1
3


논리 연산자 자체가 값을 반환하지 않고 결과가 알려지는 시점까지만 계산이 되기 때문에 최종 계산 시점의 레퍼런스 객체를 리턴하게 됩니다.

1. 1 and 2 and 3 and 4의 경우에는 4까지 모두 계산해야 하므로 최종 계산 시점인 4가 출력이 되고
2. 1 or 2 or 3 or 4의 경우 or 연산자 특성상 둘중에 하나만 참이면 모두 참인 특성에 의하여 1이란 수 자체가 True이므로 1과 2의 연산을 하지 않고 바로 1을 출력이 됩니다.
3. 1 and 2 or 3 and 4의 경우 1과 2의 연산을 하고 2 or 3의 경우 2까지 이미 True이므로 연산하지 않고 2가 출력 됩니다.
4. (1 and 2) or (3 and 4)의 경우 1 and 2 의 결과인 2와 3 and 4의 결과인 4가 2 or 4 가 되므로 2가 출력 됩니다.
5. 1 or 2 and 3 or 4의 경우 개인적인 생각으로 1 or 2는 이미 1 자체가 true이고 1 and 3에선 3이 출력되고 3 or 4에서 3이 출력되므로 3이 될것 같았지만 결국 1이 출력 되었는데 아마 1 or 2에서 1 자체가 true이므로 뒤에 연산은 보지 않고 출력되는거 같습니다
6. (1 or 2) and (3 or 4)의 경우 1 or 2는 1, 3 or 4 는 3 이므로 1 and 3은 3이므로 3이 출력 되겠습니다.

###•키보드로 정수값을 입력 받고 그 값이 양수인지 음수인지를 비트 연산자를 이용하여 판단하는 프로그램을 작성하시오.

In [9]:
number = input("input : ")

print bin(number)
print number.bit_length()

if(number >> number.bit_length()) == 0:
    print "입력한 수 (" + str(number) +")는 양수 입니다."
else:
    print "입력한 수 (" + str(number) +")는 음수 입니다."

input : 17
0b10001
5
입력한 수 (17)는 양수 입니다.


number에 수를 입력받아 number의 최상위 비트가 0인이 1인지 판별하여 양수 음수를 판단하였습니다.
그렇게 하기 위해서 number의 비트길이를 알아야 하는데 number.bit_length()을 통하여 찾았고
number를 17이라 입력받고 비트수를 찾으면 5가 나오는데 이때 이 수를 2진수로 표현해보면 0b10001이 나오는데 -17을 2진수로 표현하면 -0b10001이 나온다. 확실히 이렇게 해서는 잘 모르겠으나 아마 최상위 비트는 음수 양수를 판단하는 비트라 가정하고 number를 number의 비트만큼 오른쪽으로 밀어버리면 부호비트만 남을거라 생각하고 짜본 결과 생각대로 부호만 남아 음수 양수를 판단할수 있었습니다.

###•경로에 해당하는 문자열 1개를 입력 받아 그 안에 디렉토리 경로명과 파일명을 분리하여 리스트로 반환하는 함수 div_path(s)를 작성하시오.
◾인자로 전달하는 문자열은 경로만 들어간다고 가정한다.
◾각 디렉토리와 파일을 구분하는 문자는 '/'로 가정한다.
◾반환하는 리스트의 첫번째 원소는 디렉토리이고 두번째 원소는 파일명이다.
◾다음과 같은 실행 및 출력 결과가 도출되어야 한다.◦div_path('/usr/local/bin/python')◦['/usr/local/bin', 'python']
◦div_path('/home/chulsoo/test.txt')◦['/home/chulsoo', 'test.txt']




In [2]:
path = raw_input("input path :")

def div_path(input_path):
    index = input_path.rindex("/")
    input_path = [input_path[0:index], input_path[index+1:]]
    return input_path
    
print div_path(path)

input path :/home/usr/iejfd.txt
['/home/usr', 'iejfd.txt']


먼저 입력받은 경로에서 오른쪽부터 "/"를 찾습니다. 어차피 제일 먼저 나온 인덱스값만 나오므로 그 값을 기점으로 디렉토리 경로와 파일명을 분리시켜 input_path를 리스트 형태로 두 값을 넣었습니다.

###•두 개의 리스트를 인자로 받아서 그 두 개의 리스트에 대한 '합집합'을 반환하는 함수 list_union(lista, listb)를 작성하시오.
◾인자로 전달하는 리스트 2 개에는 정수값만 들어간다고 가정하자.
◾함수 내에서 새로운 리스트를 만들어 그 리스트 내에 인자로 받은 두 리스트의 모든 원소를 넣어 반환한다.
◾반환하는 리스트에는 절대로 중복된 원소가 들어 있으면 안된다 (집합의 조건).
◾반환하는 리스트는 정렬이 되어 있어야 한다.
◾다음과 같은 실행 및 출력 결과가 도출되어야 한다.◦list_union([1, 2, 3], [1, 2, 4])◦[1, 2, 3, 4]

◦list_union([-10, -5, 0, -1], [100, 9, 0, 9])◦[-10, -5, -1, 0, 9, 100]

◦list_union([0, 1, 2], [0, 1, 2])◦[0, 1, 2]


◾[참고] 리스트(l) 내에 새로운 정수값 (예를 들어 10)을 넣는 방법은 l.append(10) 이다.
◾[참고] 임의의 정수값 (x)이 리스트 (l) 내에 존재하는지 판단하는 방법은 x in l 이다.


In [24]:
a = [-10, -5, 0, -1]
b = [100, 9, 0, 9]

def list_union(lista, listb):
    i = lista
    for x in listb:
        if not(x in i):
            i.append(x)
    i.sort()
    return i

print list_union(a, b)

[-10, -5, -1, 0, 9, 100]


a리스트와 b리스트를 만들어 함수에 넣는 형태로 만들었습니다.
i리스트를 a리스트의 정보를 가지고
for문을 이용하여 listb에 있는 원소들을 i리스트와 비교하여 중복되는 부분을 제거하고 마지막에 기본 메소드인 sort()를 하여 정렬을 하고 i리스트를 리턴하는 형태로 작성하였습니다.

### Problem 4

앞에서부터 읽을 때나 뒤에서부터 읽을 때나 모양이 같은 수를 대칭수(palindrome)라고 부릅니다.

두 자리 수를 곱해 만들 수 있는 대칭수 중 가장 큰 수는 9009 (= 91 × 99) 입니다.

세 자리 수를 곱해 만들 수 있는 가장 큰 대칭수는 얼마입니까?

In [39]:
result = 0
first = 0
second = 0

for x in range(999,1,-1):
    for y in range(999,1,-1):
        temp = x*y
        if str(temp)[0:1] == str(temp)[-1:] and \
        str(temp)[1:2] == str(temp)[-2:-1] and \
        str(temp)[2:3] == str(temp)[-3:-2]:
            if result < temp:
                first = x
                second = y
                result = temp
        

print "가장 큰 대칭수는 " + str(first) + "x" + str(second) + "=" + str(result) + "입니다."

가장 큰 대칭수는 993x913=906609입니다.


가장 큰 대칭수를 찾기 위해서 999 x 999부터 1씩 줄여 나가며 곱하여 결과값을 문자열로 만들어 대칭되는 수끼리 같은지 확인하여 제일 큰 수를 반영하며 결과값을 찾는 방법을 사용하였습니다.

###Problem 5
1 ~ 10 사이의 어떤 수로도 나누어 떨어지는 가장 작은 수는 2520입니다.

그러면 1 ~ 20 사이의 어떤 수로도 나누어 떨어지는 가장 작은 수는 얼마입니까?

In [36]:
number = input("Input Number : ")
numlist = []
resultnumber = 1

def Maxmi(numberlist):
    finded = 0
    for item in range(2, numberlist):
        if numberlist % item == 0:
            finded = 1
            break
    if not finded:
        for x in range(2, number+1):
            multiple = 0
            while 1:
                if x%numberlist == 0 :
                    x /= numberlist
                    multiple += 1
                else :
                    exmul = 0
                    if multiple > 0 :
                        for y in numlist:
                            if y[0] == numberlist and y[1] <= multiple:
                                numlist.remove([y[0],y[1]])
                            elif y[0] == numberlist and exmul < y[1]:
                                exmul = y[1]
                        if multiple > exmul:
                            numlist.append([numberlist, multiple])
                            exmul = 0
                    break

for findnum in range(2, number+1):
    Maxmi(findnum)

for result in numlist:
    resultnumber *= (result[0] ** result[1])
    
print resultnumber
    

Input Number : 20
232792560


이 문제는 소인수분해를 이용하면 쉽게 알 수 있습니다.
1부터 20까지 수를 소인수분해를 하면
소수 2, 3, 5, 7, 11, 13, 17, 19가 나오고
이 소수들이 1~20 중에 최대 몇 번 곱해졌는지 찾고 이 찾은 수들을 모두 곱하면 답이 나옵니다.
즉, 2는 16에서 4번 곱해졌고, 3은 9와 18에서 2번 곱해졌고, 나머지 소수들은 각각 1번씩 들어가므로
16 x 6 x 5 x 7 x 11 x 13 x 17 x 19 가 답이 나옵니다.

###Problem 6
1부터 10까지 자연수를 각각 제곱해 더하면 다음과 같습니다 (제곱의 합).

1^2 + 2^2 + ... + 10^2 = 385

1부터 10을 먼저 더한 다음에 그 결과를 제곱하면 다음과 같습니다 (합의 제곱).

(1 + 2 + ... + 10)^2 = 552 = 3025

따라서 1부터 10까지 자연수에 대해 "합의 제곱"과 "제곱의 합" 의 차이는 3025 - 385 = 2640 이 됩니다.

그러면 1부터 100까지 자연수에 대해 "합의 제곱"과 "제곱의 합"의 차이는 얼마입니까?

In [29]:
a=0; b=0

for x in range(1, 101):
    a += x ** 2
    b += x

b = b ** 2

print b-a

25164150


이 문제는 for문을 이용하여 a의 제곱의 합과 b의 합을 하여 최종적으로 b을 제곱시켜 둘의 차이를 알 수 있는 문제입니다.
결과대로 25164150이 나옵니다.

###[Incremental Project] 
웹 URL로 지정된 웹페이지를 문자열로 가져와 1) 모든 HTML 태그를 제외한 순수 텍스트 문자열만을 걸러내고, 2) 그 순수 텍스트 문자열 안에 존재하는 단어가 총 몇개인지 출력하는 프로그램을 작성하시오.

In [38]:
import urllib2

source = urllib2.urlopen("http://cse.kut.ac.kr/").read()

line = ''
flag = 1

for x in source:
    if x == "<" and flag == 1:
        flag = 0
    elif x == ">" and flag == 0:
        flag = 1
        
    if x != ">" and flag == 1:
        line += x

line = line.split()

for x in line:
    print x

print "총 단어 갯수는 : " + str(len(line)) + "개 입니다."


컴퓨터공학부
//
Schoolfun
-
Study
and
Fun
--
로그인
로그인
아이디
비밀번호
로그인
유지
브라우저를
닫더라도
로그인이
계속
유지될
수
있습니다.
로그인
유지
기능을
사용할
경우
다음
접속부터는
로그인할
필요가
없습니다.
단,
게임방,
학교
등
공공장소에서
이용
시
개인정보가
유출될
수
있으니
꼭
로그아웃을
해주세요.
회원가입
ID/PW
찾기
IT
교육의
글로벌
메카!한국기술교육대학교
컴퓨터공학부
Close
소개
구성원소개
학사안내
입학
학부
대학원
학부광장
자료실
대학원
공학교육인증
Registration
컴퓨터공학부
School
of
Computer
Science
and
Engineering
IT
융합소프트웨어
인력양성사업
CK
(Creative
Korea)-I
교육역량강화사업
온라인
컨텐츠
Online
Contents
for
Education
Capability
Empowerment
IT
교육의
글로벌
메카
한국기술교육대학교
컴퓨터공학부
홈페이지에
오신
것을환영합니다!
IT
융합소프트웨어
인력양성사업
글로벌
수준의
IT전문인력
양성
IT융합분야에
대한
국제적인
경쟁력
고취
바로가기
교육역량강화사업
컨텐츠
교육역량강화사업
온라인
컨텐츠를
소개합니다.
바로가기
공지사항
[특성화]
토익,
오픽
강좌
수요조사
공고
2015-10-06
16:41
16:41
[특성화]
IT융합
개도국
해외기술지원봉사
참가자
모집...
2015-10-05
16:38
16:38
[특성화]
장비대여
안내
2015-10-05
15:28
15:28
제4회
TOPCIT
정기평가
2015-10-05
14:56
14:56
[특성화]
국제게임전시회
G-star
2015
참관
안내
2015-10-05
11:51
11:51
자유게시판
(주)
엔에스이
에서
개발자를
모집합니다.
2015-09-30
13:30
13:30
졸업작품
전시회
컴퓨터공학부
인기상
투표
이벤트
2015-09-25
01:21
01:21
삼성
대학생
프로그래밍
경진대회
2015-09-15
11:17
11:17
2

마지막 문제는 이렇게 하는것이 맞는지 모르겠네요
<와 >를 기점으로 문자열로 나열하여 저장한 후 split를 이용하여 공백단위로 잘라봤는데 이것을 그대로 출력할땐 아스키값이 나와서 의아 했는데
반복문을 사용하여 출력하면 제대로 나오는걸 볼 수 있습니다. 그렇게 하여 리스트의 갯수를 단어의 갯수로 하여 출력 하였습니다.

split를 사용하면 아스키값이 나와서 좀 난감한점도 있었지만 이것을 하나하나 반복문으로 출력할땐 한글이 제대로 나와서 무엇이 잘못되었는지 잘 모르겠네요 이번 과제는 오일러문제 5번이 약간 어려운거 말고는 무난했던거 같습니다.